files/grest/rpc/01_cached_tables/stake_snapshot_cache.sql

```diff
@@ -0,0 +1,445 @@
+/* Keeps track of stake snapshots taken at the end of epochs n - 1 and n - 2 */
+DROP TABLE IF EXISTS GREST.stake_snapshot_cache;
+
+CREATE TABLE GREST.stake_snapshot_cache (
+  STAKE_ADDRESS varchar,
+  POOL_ID varchar,
+  AMOUNT numeric,
+  EPOCH_NO bigint,
+  PRIMARY KEY (STAKE_ADDRESS, EPOCH_NO)
+);
+
+-- This function captures the stake snapshot of previous epoch
+CREATE OR REPLACE PROCEDURE GREST.CAPTURE_LAST_EPOCH_SNAPSHOT ()
+LANGUAGE PLPGSQL
+AS $$
+DECLARE
+  _previous_epoch_no bigint;
+  _previous_epoch_last_block_id bigint;
+  _active_stake_baseline_epoch bigint;
+  _last_active_stake_block_id bigint;
+  _lower_bound_account_tx_id bigint;
+  _upper_bound_account_tx_id bigint;
+BEGIN
+  IF (
+    -- If checking query with the same name there will be 2 results
+    SELECT COUNT(pid) > 1
+      FROM pg_stat_activity
+      WHERE state = 'active'
+        AND query ILIKE '%GREST.CAPTURE_LAST_EPOCH_SNAPSHOT(%'
+        AND datname = (
+          SELECT current_database()
+        )
+  ) THEN
+      RAISE EXCEPTION 'Previous query still running but should have completed! Exiting...';
+  END IF;
+
+  IF EXISTS (
+    SELECT FROM grest.stake_snapshot_cache
+      WHERE epoch_no = (SELECT MAX(NO) - 1 FROM PUBLIC.EPOCH)
+        LIMIT 1
+  ) THEN
+    RETURN;
+  END IF;
+
+  SELECT MAX(NO) - 1 INTO _previous_epoch_no FROM PUBLIC.EPOCH;
+
+  SELECT _previous_epoch_no - 2 INTO _active_stake_baseline_epoch;
+
+  -- Set-up interval limits for previous epoch
+  SELECT id INTO _last_active_stake_block_id FROM PUBLIC.BLOCK
+    WHERE epoch_no = _active_stake_baseline_epoch
+      AND block_no IS NOT NULL
+    ORDER BY block_no DESC LIMIT 1;
+
+  SELECT id INTO _previous_epoch_last_block_id FROM PUBLIC.BLOCK
+    WHERE epoch_no = _previous_epoch_no
+      AND block_no IS NOT NULL
+    ORDER BY block_no DESC LIMIT 1;
+
+  SELECT MAX(id) INTO _lower_bound_account_tx_id FROM PUBLIC.TX
+    WHERE block_id <= _last_active_stake_block_id;
+
+  SELECT MAX(id) INTO _upper_bound_account_tx_id FROM PUBLIC.TX
+    WHERE block_id <= _previous_epoch_last_block_id;
+
+/* Registered and delegated accounts to be captured (have epoch_stake entries for baseline) */
+  WITH
+    latest_non_cancelled_pool_retire as (
```

Comment:
Both `latest_non_cancelled_pool_retire`, `minimum_pool_delegation_tx_ids` CTEs could explore using `GROUP BY` instead of `SELECT DISTINCT`.